src/Juvix/FrontendContextualise/ModuleOpen/Environment.hs

```diff
@@ -100,3 +169,205 @@ lookupModMap s =
+                    grabList new
+                  NameSpace.List {publicL = pubO} =
+                    grabList old
+               in fmap fst pubN <> fmap fst pubO
+
+-- we don't take a module mapping as all symbols at the point of
+-- resolve can't be anything else, thus we don't have to pre-pend any
+-- qualification at this point... we add qualification later as we add
+-- things, as we have to figure out the full resolution of all opens
+
+-- | @resolve@ takes a context and a list of open modules and the modules
+-- they are open in (the module itself, and sub modules, which the opens are
+-- implicit), and fully resolves the opens to their full name.
+-- for example @open Prelude@ in the module Londo translates to
+-- @resolve ctx PreQualified {opens = [Prelude]}@
+-- == @Right (fromList [(TopLevel :| [Londo],[Explicit (TopLevel :| [Prelude])])])@
+resolve :: Context.T a b c -> [PreQualified] -> Either Error OpenMap
+resolve ctx = foldM (resolveSingle ctx) mempty
+
+resolveSingle ::
+  Context.T a b c -> OpenMap -> PreQualified -> Either Error OpenMap
+resolveSingle ctx openMap Pre {opens, implicitInner, explicitModule} =
+  case Context.switchNameSpace explicitModule ctx of
+    Left (Context.VariableShared err) ->
+      Left (IllegalModuleSwitch err)
+    Right ctx -> do
+      resolved <- pathsCanBeResolved ctx opens
+      qualifiedNames <- resolveLoop ctx mempty resolved
+      openMap
+        |> Map.insert explicitModule (fmap Explicit qualifiedNames)
+        |> ( \map ->
+               foldr
+                 (\mod -> Map.insert mod (fmap Implicit qualifiedNames))
+                 map
+                 implicitInner
+           )
+        |> Right
+
+-- this goes off after the checks have passed regarding if the paths
+-- are even possible to resolve
+resolveLoop ::
+  Context.T a b c -> ModuleMap -> Resolve a b c -> Either Error [NameSymbol.T]
+resolveLoop ctx map Res {resolved, notResolved = cantResolveNow} = do
+  map <- foldM addToModMap map fullyQualifyRes
+  --
+  let newResolve = resolveWhatWeCan ctx (qualifyCant map <$> cantResolveNow)
+  --
+  if  | null cantResolveNow ->
+        Right qualifedAns
+      | length (notResolved newResolve) == length cantResolveNow ->
+        Left (CantResolveModules cantResolveNow)
+      | otherwise ->
+        (qualifedAns <>) <$> resolveLoop ctx map newResolve
+  where
+    fullyQualifyRes =
+      Context.resolveName ctx <$> resolved
+    qualifedAns =
+      fmap snd fullyQualifyRes
+    addToModMap map (def, sym) =
+      case def of
+        Context.Record cont _ ->
+          let NameSpace.List {publicL} = NameSpace.toList cont
+           in foldlM
+                ( \map (x, _) ->
+                    case map Map.!? x of
+                      -- this means we already have opened this in another
+                      -- module
+                      Just _ -> Left (AmbiguousSymbol x)
+                      -- if it's already in scope from below or above us
```

Comment:
hmm, will this result in things being in scope with the names the programmer expects?

Reply:
Which part? 

If the symbol is already in the map, then that means `Open A` has symbol `x`. So when we get to `Open B` which also has symbol `x` we should have an ambigious open.

Reply:
ah I see